### PR TITLE
gtk patch backports

### PR DESCRIPTION
**Pull request by amaxine**
###### Description of changes

cc @jtojnar 

The gtk 3 hash is just the icons fix + some translation updates.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/gtk/3.x.nix [hunk 1]
```diff
@@ -71,8 +71,8 @@ stdenv.mkDerivation rec {
   ];
 
   src = fetchurl {
-    url = "mirror://gnome/sources/gtk+/${lib.versions.majorMinor version}/gtk+-${version}.tar.xz";
-    sha256 = "sha256-WIsGUi4l0VeemJtvnYob2/L+E83gGgTpBP80aiJeeAE=";
+    url = "https://gitlab.gnome.org/GNOME/gtk/-/archive/9d1d2f0a6643570274121fc1473e46a6edc2e32d/gtk-9d1d2f0a6643570274121fc1473e46a6edc2e32d.tar.gz";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/3.x.nix:74]: This is an auto-generated archive so it needs to be fetched using `fetchzip` (or `fetchgit`).

File: pkgs/development/libraries/gtk/4.x.nix [hunk 2]
```diff
@@ -77,6 +78,14 @@ stdenv.mkDerivation rec {
     sha256 = "pXrNDkSCmBcA/fhllsdBPLYe9H915HR/2oCegjG42Ww=";
   };
 
+  patchFlags = ["--binary"];
+  patches = [
+    (fetchpatch {
+      url = "https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4670.patch";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:84]:
```suggestion
      url = "https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4670.patch";
```
```suggestion
    # Fix GNOME Control Center crash.
    # https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4670
    (fetchpatch {
      url = "https://gitlab.gnome.org/GNOME/gtk/-/commit/ae0166973795e750f08b89f9f0ef974d7ac48bc7.patch";
```

File: pkgs/development/libraries/gtk/4.x.nix [hunk 3]
```diff
@@ -77,6 +78,14 @@ stdenv.mkDerivation rec {
     sha256 = "pXrNDkSCmBcA/fhllsdBPLYe9H915HR/2oCegjG42Ww=";
   };
 
+  patchFlags = ["--binary"];
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:81]: What is this for?

Comment (maxeaubrey) [reply pkgs/development/libraries/gtk/4.x.nix:81]: Absolutely nothing, me being dumb and doing this late at night the last two days :D

**Comment by maxeaubrey** [comment]
Fixed up the changes. Thanks for the quick review.

**Review by jtojnar — APPROVED** [review]

**Comment by jtojnar** [comment]
GTK changes https://gitlab.gnome.org/GNOME/gtk/-/compare/3.24.33...9d1d2f0a6643570274121fc1473e46a6edc2e32d

**Comment by Mindavi** [comment]
I think the commit 5d15a099b94be16ee619053ed6bcbddefd561a19 broke staging-next gtk4 builds with the following error:

```
unpacking sources
unpacking source archive /nix/store/s5d4w28f6an5kc7ddaiszs03qpjb9qkc-gtk-4.6.4.tar.xz
source root is gtk-4.6.4
setting SOURCE_DATE_EPOCH to timestamp 1652464903 of file gtk-4.6.4/gtk/theme/Default/Default-hc-dark.css
patching sources
applying patch /nix/store/9kb4143hg292rihq15pjly3m2xgvq1f9-ae0166973795e750f08b89f9f0ef974d7ac48bc7.patch
patching file gtk/gtkboxlayout.c
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
3 out of 3 hunks ignored -- saving rejects to file gtk/gtkboxlayout.c.rej
```

Probably because gtk 4.6.4 was released in between, causing this patch not to apply anymore.

**Comment by jtojnar** [comment]
https://github.com/NixOS/nixpkgs/commit/ff29d2caa935a8e6216380dfcd7ee714fd8bd5bf removes the patch. Perhaps master was not merged correctly into staging-next?

**Comment by Mindavi** [comment]
Hmm yeah, maybe that's the issue. Not sure why that happened. But definitely visible in the diff as well: https://github.com/NixOS/nixpkgs/compare/master...staging-next

**Pull request merged**